better 3D rendering

ffreyer opened the pull request:
So far I've tried updating the mesh implementation, which was somewhat successful. One example renders well (thought the file is quite large) [test.pdf](https://github.com/JuliaPlots/CairoMakie.jl/files/5180139/test.pdf), while others don't render at all. For example, neither the cat mesh nor a simple `Sphere(Point3f0(0), 1f0)` renders and I have no idea why...

The model is from [learnopengl.com](https://github.com/JoeyDeVries/LearnOpenGL/tree/master/resources/objects/nanosuit).

ffreyer commented:
It's text rendering that makes this fail. Skipping text rendering allows me to render [smudge-cat](https://github.com/JuliaPlots/CairoMakie.jl/files/5180325/test.pdf)

ffreyer commented:
I don't know what I'm doing but Axis3D text renders somewhat reasonably now. Not sure about text in general though
[axis3d_test.pdf](https://github.com/JuliaPlots/CairoMakie.jl/files/5185178/axis3d_test.pdf)

ffreyer commented:
TODO:
- [ ] back face culling
- [ ] surface
- [ ] meshscatter
- [ ] countour?

SimonDanisch commented:
Cool! :) Maybe comes a bit late,but did you see: https://github.com/JuliaPlots/AbstractPlotting.jl/tree/master/src/jl_rasterizer ?

ffreyer commented:
I'm not creating raster images here so I didn't think it would be useful to look at. Is there anything I should check out in particular?

SimonDanisch commented:
Well, it's also shading triangles and tried to make the shaders swappable etc..

ffreyer commented:
Afaik you can only really interpolate colors on triangles with Cairo, so I figured there isn't much use to anything beyond basic shading.

SimonDanisch commented:
Fair enough ;)

ffreyer commented on src/primitives.jl:
```diff
@@ -462,3 +491,161 @@ function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlott
+#                                 MeshScatter                                  #
+################################################################################
+
+function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlotting.MeshScatter)
+    m = normal_mesh(primitive[:marker][])
+    pos = primitive[1][]
+    # For correct z-ordering we need to be in view/camera or screen space
+    model = primitive[:model][]
+    view = scene.camera.view[]
+    sort!(pos, by = p -> begin
+        p4d = to_ndim(Vec4f0, to_ndim(Vec3f0, p, 0f0), 1f0)
+        cam_pos = view * model * p4d
+        cam_pos[3] / cam_pos[4]
+    end, rev=false)
+
+    colors = primitive[:color][]
+
+    for i in eachindex(pos)
+        p = pos[i]
+        if colors isa Vector
+            primitive[:color][] = colors[i]
+        end
+        draw_mesh3D(scene, screen, primitive, m, p, primitive[:markersize][])
+    end
+
+    primitive[:color][] = colors
+    return nothing
+end
```
I thought about creating new `Mesh` plots here (and in the method for surface) since that's the primitive I'm drawing in the end. What would be the way to do that? 

I tried copying stuff directly, i.e. `Mesh(parent, transformation, attributes, input_args, converted)`, but that needs type information (`Mesh{...}`) which I haven't figure out yet.

SimonDanisch replied:
The loop seems pretty fair?  I'm not sure if it would be any easier to create a mesh, which would also need a very similar loop?

ffreyer replied:
I don't feel good about modifying attributes of the input. I just have `:color` here, but I assume there are few more that I'm missing. I could also pass all of those to `draw_mesh3D ` directly, which would probably be the easiest solution.

Creating a new Mesh object seems like the most readable/conceptually straight forward to me. `Surface` and `MeshScatter` objects would be transformed to `Mesh` objects and then just drawn without any extra fuzz. But maybe that's just me getting pedantic...

ffreyer commented:
I think I'm done with this. `meshscatter` works including rotations and markersize, `surface` works and `mesh` works. I added a `faceculling` attribute, because it seems to sometimes mess up. You can essentially set the threshold with it which should be `<= 0.0` (anything `< -1` would mean no culling). I also added `specular` lighting.

![Screenshot from 2020-09-19 00-30-01](https://user-images.githubusercontent.com/10947937/93653308-9be72f80-fa18-11ea-95f7-715c5259c652.png)
![Screenshot from 2020-09-19 01-37-46](https://user-images.githubusercontent.com/10947937/93653340-c2a56600-fa18-11ea-8238-87d42cbc447c.png)

ffreyer commented:
I looked through some of the current test failures - most of them use `rand` or `randn`. One exception I found is http://juliaplots.org/MakieReferenceImages/gallery//rotation_1/index.html which has the wrong axis. (I'm on AbstractPlotting master)

SimonDanisch commented:
Thanks a lot! I incorperated the changes in https://github.com/JuliaPlots/CairoMakie.jl/pull/109
